cmd/metal-api/internal/datastore/image.go

```diff
@@ -35,3 +47,146 @@ func (rs *RethinkStore) DeleteImage(i *metal.Image) error {
+		if err != nil {
+			return nil, err
+		}
+		images = is
+	}
+	if machines == nil {
+		ms, err := rs.ListMachines()
+		if err != nil {
+			return nil, err
+		}
+		machines = ms
+	}
+	firstOSImage := make(map[string]bool)
+	result := metal.Images{}
+	sortedImages := sortImages(images)
+	for _, image := range sortedImages {
+		// Always keep the most recent image for one OS even if no machine uses it is not valid anymore
+		// this prevents that there is no image at all left if no new images are pushed.
+		_, ok := firstOSImage[image.OS]
+		if !ok {
+			firstOSImage[image.OS] = true
+			continue
+		}
+
+		if isOrphanImage(image, machines) {
+			err := rs.DeleteImage(&image)
+			if err != nil {
+				return nil, fmt.Errorf("unable to delete image:%s err:%v", image.ID, err)
+			}
+			result = append(result, image)
+		}
+	}
+	return result, nil
+}
+
+// isOrphanImage check if a image is not allocated and older than allowed.
+func isOrphanImage(image metal.Image, machines metal.Machines) bool {
+	if time.Since(image.ExpirationDate) < 0 {
+		return false
+	}
+	orphan := true
+	for _, m := range machines {
+		if m.Allocation == nil {
+			continue
+		}
+		if image.ID == m.Allocation.ImageID {
+			return false
+		}
+	}
+	return orphan
+}
+
+// getMostRecentImageFor
+// the id is in the form of: <name>-<version>
+// where name is for example ubuntu or firewall
+// version must be a semantic version, see https://semver.org/
+// we decided to specify the version in the form of major.minor.patch,
+// where patch is in the form of YYYYMMDD
+// If version is not fully specified, e.g. ubuntu-19.10 or ubuntu-19.10
+// then the most recent ubuntu image (ubuntu-19.10.20200407) is returned
+// If patch is specified e.g. ubuntu-20.04.20200502 then this exact image is searched.
+func (rs *RethinkStore) getMostRecentImageFor(id string, images metal.Images) (*metal.Image, error) {
+	os, sv, err := rs.GetOsAndSemver(id)
+	if err != nil {
+		return nil, err
+	}
+
+	matcher := "~"
+	// if patch is given return a exact match
+	if sv.Patch() > 0 {
+		matcher = "="
+	}
+	constraint, err := semver.NewConstraint(matcher + sv.String())
+	if err != nil {
+		return nil, fmt.Errorf("could not create constraint of image version:%s err:%w", sv, err)
+	}
+
+	var latestImage *metal.Image
+	sortedImages := sortImages(images)
+	for _, image := range sortedImages {
+		if !strings.HasPrefix(id, image.OS) {
+			continue
+		}
+		if time.Since(image.ExpirationDate) > 0 {
+			continue
+		}
+		v, err := semver.NewVersion(image.Version)
+		if err != nil {
+			continue
+		}
+		if constraint.Check(v) {
+			latestImage = &image
+			break
+		}
+	}
+	if latestImage != nil {
+		return latestImage, nil
+	}
+	return nil, fmt.Errorf("no image for os:%s version:%s found", os, sv)
+}
+
+// GetOsAndSemver parses a imageID to OS and Semver, or returns an error
+func (rs *RethinkStore) GetOsAndSemver(id string) (string, *semver.Version, error) {
+	imageParts := strings.Split(id, "-")
```

Comment:
only cosmetic, but i think a `strings.SplitN(id, "-", 2)` would work too and you do not need to join the version-parts later.

Reply:
pointed out a inconsequence and missing test, modified and clarified with test